merry-go-round.md

```diff
@@ -0,0 +1,133 @@
+
+The Ethereum state is however, constantly changing as each new block is added
+to the chain.  The result is that as a node walks through the network
+requesting state data, the first proofs they receive will be anchored to some
+state root that we'll call `S_n` and at some point, the proofs they receive will
+be anchored to some subsequent root that we'll call `S_n+1`.
+
+### Dealing with updates to the state
+
+* Problem: As the Ethereum state changes, nodes that are syncing the state must
+  be able to *update* the data they have already received to newer and newer
+  state roots.
+
+To address this, we will introduce two changes to how nodes in the network
+handle state.
+
+First, we introduce a concept of epoch boundaries.  Nodes in the network will
+store proof of the segment of the state that they are storing for the network
+at the most recent state root from the most recent epoch boundary.  The result
+of this is that within a single epoch, all requests for state data from
+different nodes will be anchored to the same state root.
+
+Under this model we still assume that all updates to the state are being
+gossiped as each block is added to the chain.  Nodes that are storing data
+would accumulate a state diff from these proofs during the course of an epoch,
+and at the end of the epoch, they can apply this state diff to their stored
+state data to update their stored state to the new state root at the new epoch
+boundary.
+
+Research should be done to determine optimal epoch lengths.  The epoch length
+must be short enough that a node can fully traverse the full DHT within a
+single epoch.  We suggest 256 (51 minutes) as a starting point.
+
+## Specification
+
+### Storage
+
+Nodes in the network store two things.
+
+* proofs about the state *near* their position in the DHT anchored at the most
+  recent epoch boundary
+* diffs about the state accumulated from gossip that allow updating the state
+  from the most recent epoch boundary to the newest state root.
+
+> Note, that nodes will likely need to store at least the most recent two
+> epochs of state data to account for the points in time where we are
+> transitioning between epochs.  Nodes should continue to be able to retrieve
+> data from the previous epoch to allow them the necessary time to fully obtain
+> the necessary proof data to update their view of the state to the newest
+> epoch.
+
+### Gossip
+
+The network provides a gossip mechanism that disseminates new state data from
+each block to the nodes in the network that would store that data.
+
+### Retrieval
+
+The network exposes request/response primatives for:
+
+* retrieving a node's proof of the state from a recent epoch boundary, likely bounded by a range of keys within the trie.
+* retrieving a a node's proof diff that allows you to update their state proof to either a newer state root or the latest epoch boundary.
+
+
+### Syncing
+
+The process for syncing then becomes the following.
```

Comment:
This is maybe just a matter of taste:

I only mostly understood process, but I think I would have understood it just as well in a much more compact form without the lettered definitions. A deeper explanation is valuable, but I don't think this accomplishes it. So probably it would be helpful to have two versions in a row: a shorter version that would help give the overview, and then a deeper version of it that would fill in the gaps.

I think I'm missing some details of what happens at the MGR epoch boundaries, for example. Like, what happens if you haven't collected all the base state by the next epoch boundary? Do you start over?

Reply:
Good feedback, I generally knocked this out in one go with the knowledge that it'd need a lot of revising to make it clear to read.  I'll give this another iteration soon.

Reply:
> I think I'm missing some details of what happens at the MGR epoch boundaries, for example. Like, what happens if you haven't collected all the base state by the next epoch boundary? Do you start over?

I *think* that at each epoch boundary you have `P_n` which is the state data you were able to gather within the epoch, and then you start building `P_n+1` in the new epoch starting wherever  `P_n` finished.  Once you've aquired the diff data for `P_n` you merge it with `P_n+1`.  I think the only "race" here is ensuring that acquisition of the diff data can be reliably done within the time frame of a single epoch **or** ensuring that clients hold on to enough recent epochs... which are effectively the same thing.  The only *bad* case here is a client that is too slow to acquire the diff before the epoch the proof is based on gets purged from the communal storage.